### PR TITLE
kdesu: provide a script in bin and look for daemon first in /run/wrappers/bin

### LINKED ISSUE #28426: Installing kdesu
When setting advanced properties on a program/link to start as another user it is throwing the error:

    Sorry - Plasma
    KDEInit could not launch 'kdesu':
    Could not open library 'libkdeinit5_kdesu'.
    Cannot load library /run/current-system/sw/lib/libkdeinit5_kdesu: (/run/current-system/sw/lib/libkdeinit5_kdesu.so: cannot open shared object file: No such file or directory)

I tried to install kdesu using:

    [root@nixos:~]# nix-env -qaP kdesu
    nixos.kdeFrameworks.kdesu  kdesu-5.36.0

    [root@nixos:~]# nix-env -iA nixos.kdeFrameworks.kdesu
    replacing old ‘kdesu-5.36.0’
    installing ‘kdesu-5.36.0’
    building path(s) ‘/nix/store/2bn27h5drw0800gk5z2rmwfx78gqbbdc-user-environment’
    created 2209 symlinks in user environment

    [root@nixos:~]# kdesu
    bash: kdesu: command not found

    [root@nixos:~]# which kdesu
    which: no kdesu in (/root/bin:/run/wrappers/bin:/etc/per-user-pkgs/root/bin:/root/.nix-profile/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin)

I also tried installing `nix-env -iA nixos.kde-cli-tools` but that does not work either.

Any suggestions on how to get kdesu installed on NixOS?

Environment:

    [root@nixos:~]# nixos-version
    17.09pre113100.d25a9f57e3 (Hummingbird)

    [root@nixos:~]# nix-env --version
    nix-env (Nix) 1.11.13

    [root@nixos:~]# nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion
    "17.09pre113100.d25a9f57e3"

    [root@nixos:~]# grep build-use-sandbox /etc/nix/nix.conf
    build-use-sandbox = false

Stack Overflow link: https://stackoverflow.com/q/45786752/750216

### PR DESCRIPTION
**Pull request by romildo**
###### Motivation for this change

If looking for the `kdesud` first in `libexec`, the eventually wrapped one in `/run/wrappers/bin` can not be found. With this change it becomes possible to use a wrapped version of the daemon.

Also a shell script is installed at ${plasma-5.kde-cli-tools}/bin/` to ease launching `kdesu`.

Fixes https://github.com/NixOS/nixpkgs/issues/28426.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by romildo** [comment]
One still has to add something like the following in the system configuration in order to be able to use the `kdesu` daemon.

``` nix
  security.wrappers.kdesud = {
    source = "${pkgs.kdesu}/libexec/kf5/kdesud";
    setgid = true;
  };
```

File: pkgs/development/libraries/kde-frameworks/kdesu-search-for-wrapped-daemon-first.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
+From 01af4d2a098e5819c09bca37568941dcd4b89d0b Mon Sep 17 00:00:00 2001
```
Comment (ttuegel) [inline pkgs/development/libraries/kde-frameworks/kdesu-search-for-wrapped-daemon-first.patch:1]: Please create a directory for `kdesu` in `kde-frameworks` and move this patch there.

File: pkgs/development/libraries/kde-frameworks/kdesu.nix [hunk 2]
```diff
@@ -11,4 +11,5 @@ mkDerivation {
   buildInputs = [ kcoreaddons ki18n kpty kservice qtbase ];
   propagatedBuildInputs = [ kpty ];
   outputs = [ "out" "dev" ];
+  patches = [ ./kdesu-search-for-wrapped-daemon-first.patch ];
```
Comment (ttuegel) [inline pkgs/development/libraries/kde-frameworks/kdesu.nix:14]: Please move this file to `.../kde-frameworks/kdesu/default.nix` and update `kde-frameworks/default.nix`.

File: pkgs/desktops/plasma-5/kde-cli-tools.nix [hunk 3]
```diff
@@ -11,4 +11,12 @@ mkDerivation {
...
     qtx11extras kactivities plasma-workspace
   ];
+  postInstall = ''
+    cat > $out/bin/kdesu <<EOF
+    #! /bin/sh -e
+    export PATH="\''${PATH}:${kinit}/bin"
+    exec -a "\$0" $out/libexec/kf5/kdesu "\$@"
+    EOF
+    chmod +x $out/bin/kdesu
+  '';
```
Comment (ttuegel) [inline pkgs/desktops/plasma-5/kde-cli-tools.nix:21]: What is the need for this?

**Review by ttuegel — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/kde-frameworks/kdesu-search-for-wrapped-daemon-first.patch (hunk 1 above)
Comment (romildo) [reply pkgs/development/libraries/kde-frameworks/kdesu-search-for-wrapped-daemon-first.patch]: Done.

File: pkgs/development/libraries/kde-frameworks/kdesu.nix (hunk 2 above)
Comment (romildo) [reply pkgs/development/libraries/kde-frameworks/kdesu.nix:14]: Done.

File: pkgs/desktops/plasma-5/kde-cli-tools.nix (hunk 3 above)
Comment (romildo) [reply pkgs/desktops/plasma-5/kde-cli-tools.nix:21]: - have `kdesu` in `$PATH`
- avoid an error about `kinit` not found when running outside of plasma

Comment (romildo) [reply pkgs/desktops/plasma-5/kde-cli-tools.nix:21]: This is a test running `kdesu` in a VM under fluxbox. Running the original `kdesu` executable from `libexec` fails. But running the script this PR install under `bin` succeeds.

```
[guest@nixos:~]$ /nix/store/kxrmiw71jk2hs5qb65q18dyddjcxiwki-kde-cli-tools-5.17.5/libexec/kf5/kdesu xterm
kf5.kdbusaddons: Can not find 'kdeinit5' executable at  "/run/wrappers/bin:/home/guest/.nix-profile/bin:/etc/profiles/per-user/guest/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin" "/nix/store/kxrmiw71jk2hs5qb65q18dyddjcxiwki-kde-cli-tools-5.17.5/libexec/kf5, /nix/store/ymsw8j9876ivxirmyllflxgf0gpdx1d7-qtbase-5.12.7/bin"
"KLauncher could not be reached via D-Bus. Error when calling kdeinit_exec_wait:\nThe name org.kde.klauncher5 was not provided by any .service files\n"

[guest@nixos:~]$ /nix/store/kxrmiw71jk2hs5qb65q18dyddjcxiwki-kde-cli-tools-5.17.5/bin/kdesu xterm
Connecting to deprecated signal QDBusConnectionInterface::serviceOwnerChanged(QString,QString,QString)


```

Comment (romildo) [reply pkgs/desktops/plasma-5/kde-cli-tools.nix:21]: It seems that `kdeinit5` is a run time dependency that should be found in `$PATH`.

File: pkgs/desktops/plasma-5/kde-cli-tools.nix [hunk 4]
```diff
@@ -11,4 +11,12 @@ mkDerivation {
     kcmutils kconfig kdesu ki18n kiconthemes kinit kio kwindowsystem qtsvg
     qtx11extras kactivities plasma-workspace
   ];
+  postInstall = ''
+    cat > $out/bin/kdesu <<EOF
+    #! /bin/sh -e
+    export PATH="\''${PATH}:${kinit}/bin"
+    exec -a "\$0" $out/libexec/kf5/kdesu "\$@"
+    EOF
+    chmod +x $out/bin/kdesu
```
Comment (ttuegel) [inline pkgs/desktops/plasma-5/kde-cli-tools.nix:20]: The preferred way to do this is:

```suggestion
    makeWrapper $out/libexec/kf5/kdesu $out/bin/kdesu \
        --suffix PATH : ${lib.getBin kinit}/bin
```

This requires `makeWrapper` in `nativeBuildInputs`.

**Comment by romildo** [comment]
Changed to avoid double wrapping `kdesu`.

**Comment by romildo** [comment]
@ttuegel can this be merged?

**Review by ttuegel — APPROVED** [review]
Looks good to me!

**Comment by ttuegel** [comment]
Thanks! :smiley:

**Pull request merged**